tooling/nargo_fmt/src/formatter.rs

```diff
@@ -0,0 +1,303 @@
+use noirc_frontend::{
+    ast::Ident,
+    hir::resolution::errors::Span,
+    lexer::Lexer,
+    token::{Keyword, SpannedToken, Token},
+    ParsedModule,
+};
+
+use crate::Config;
+
+mod alias;
+mod attribute;
+mod chunks;
+mod comments_and_whitespace;
+mod doc_comments;
+mod expression;
+mod function;
+mod generics;
+mod global;
+mod impls;
+mod item;
+mod lvalue;
+mod module;
+mod path;
+mod pattern;
+mod statement;
+mod structs;
+mod trait_impl;
+mod traits;
+mod type_expression;
+mod types;
+mod use_tree;
+mod visibility;
+mod where_clause;
+
+pub(crate) struct Formatter<'a> {
+    config: &'a Config,
+    source: &'a str,
+    lexer: Lexer<'a>,
+    token: Token,
+    token_span: Span,
+
+    /// The current indentation level.
+    /// We let it be negative because in some cases we just want to decrease indentation
+    /// to compensate an indentation that will come later that we don't need, and we
+    /// don't want to panic when reaching those negative values.
+    indentation: i32,
```

Comment:
```suggestion
    /// We allow it to be negative because in some cases we just want to decrease indentation
    /// to preemptively cancel out an indentation that will come later which we don't want to take effect,
    /// and we don't want to panic when reaching those negative values.
    indentation: i32,
```
When I read "come later that we don't need" I thought it's something like the conditional indentation in the blog post, which either happens or not based on whether something is wrapped, and I thought it might leave some dangling negative unindent. It took me a long time to conclude that you must mean something like `-1; +1`, and not `-1; if whatever() { +1 }`. Not sure if this wording conveys that better, but wanted to highlight that I found the comment not obvious (but super helpful nevertheless!).